src/api/sfc-script-setup.md

```diff
@@ -0,0 +1,268 @@
+
+```js
+import { FooBar as FooBarChild } from './components'
+```
+
+## `defineProps` and `defineEmits`
+
+To declare options like `props` and `emits` with full type inference support, we can use the `defineProps` and `defineEmits` APIs, which are automatically available inside `<script setup>`:
+
+```html
+<script setup>
+  const props = defineProps({
+    foo: String
+  })
+
+  const emit = defineEmits(['change', 'delete'])
+  // setup code
+</script>
+```
+
+- `defineProps` and `defineEmits` are **compiler macros** only usable inside `<script setup>`. They do not need to be imported, and are compiled away when `<script setup>` is processed.
+
+- `defineProps` accepts the same value as the `props` option, while `defineEmits` accepts the same value as the `emits` option.
+
+- `defineProps` and `defineEmits` provide proper type inference based on the options passed.
+
+- The options passed to `defineProps` and `defineEmits` will be hoisted out of setup into module scope. Therefore, the options cannot reference local variables declared in setup scope. Doing so will result in a compile error. However, it _can_ reference imported bindings since they are in the module scope as well.
+
+If you are using TypeScript, it is also possible to [declare props and emits using pure type annotations](#typescript-only-features).
+
+## `defineExpose`
+
+Components using `<script setup>` are **closed by default** - i.e. the public instance of the component, which is retrieved via template refs or `$parent` chains, will **not** expose any of the bindings declared inside `<script setup>`.
+
+To explicitly expose properties in a `<script setup>` component, use the `defineExpose` compiler macro:
+
+```html
+<script setup>
+  const a = 1
+  const b = ref(2)
+
+  defineExpose({
+    a,
+    b
+  })
+</script>
+```
+
+When a parent gets an instance of this component via template refs, the retrieved instance will be of the shape `{ a: number, b: number }` (refs are automatically unwrapped just like on normal instances).
+
+## `useSlots` and `useAttrs`
+
+Usage of `slots` and `attrs` inside `<script setup>` should be relatively rare, since you can access them directly as `$slots` and `$attrs` in the template. In the rare case where you do need them, use the `useSlots` and `useAttrs` helpers respectively:
+
+```html
+<script setup>
+  import { useSlots, useAttrs } from 'vue'
+
+  const slots = useSlots()
+  const attrs = useAttrs()
+</script>
+```
+
+`useSlots` and `useAttrs` are actual runtime functions that return the equivalent of `setupContext.slots` and `setupContext.attrs`. They can be used in normal composition API functions as well.
+
+## Usage alongside normal `<script>`
+
+`<script setup>` can be used alongside normal `<script>`. A normal `<script>` maybe needed in cases where we need to:
+
+- Declare options that cannot be expressed in `<script setup>`, for example `inheritAttrs` or custom options enabled via plugins;
+- Declaring named exports;
```

Comment:
Does the `;` at the end make sense? Shouldn't it be a `.` or nothing at all in the 3 bullet points?

```suggestion
- Declare options that cannot be expressed in `<script setup>`, for example `inheritAttrs` or custom options enabled via plugins.
- Declaring named exports.
```